protocol/contracts/oud/OudRedeemer.sol

```diff
@@ -0,0 +1,145 @@
+pragma solidity ^0.8.17;
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Oud Redeemer (protocol/contracts/oud/OudRedeemer.sol)
+
+import '@openzeppelin/contracts/access/Ownable.sol';
+import '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+import '../interfaces/oud/IOudRedeemer.sol';
+import '../interfaces/oud/Oud__ITempleLineOfCredit.sol';
+import '../interfaces/oud/IOudtoken.sol';
+import '../interfaces/oud/Oud__ITempleERC20Token.sol';
+
+/**
+ * @title Allows for the redemption of Oud + a stable coin for Temple at the Temple Treasury Price index
+ */
+contract OudRedeemer is IOudRedeemer, Ownable {
+  using SafeERC20 for IERC20;
+  error AddressZero();
+  error RedeemAmountZero();
+
+  // @notice Temple token address
+  address public immutable templeToken;
+
+  // @notice Oud token address
+  address public immutable oudToken;
+
+  // @notice Stable token address
+  IERC20 public stable;
+
+  //   @notice The Temple Price Index, used to calculate amount of 'stable' required to mint Temple
+  uint256 public tpi;
+
+  // @notice The decimal precision of 'tpi'/Temple Price index
+  uint256 public constant tpiDecimals = 4;
+
+  // Target address for 'stable' token used for redemption of Oud + stable for Temple
+  address public depositStableTo;
+
+  // @dev requires this contract to have minting rights for both Temple and Oud
+  constructor(
+    address _oudToken,
+    address _templeToken,
+    address _stable,
+    address _depositStableTo,
+    uint256 _tpi
+  ) {
+    templeToken = _templeToken;
+    oudToken = _oudToken;
+    stable = IERC20(_stable);
+    depositStableTo = _depositStableTo;
+    tpi = _tpi;
+  }
+
+  //  @notice Returns the index at which 'stable' token is required to mint Temple
+  function treasuryPriceIndex() external view returns (uint256) {
+    return tpi;
+  }
+
+  // @dev Allows the tpi to be adjusted.
+  // @param 'value' is the new TPI
+  function setTreasuryPriceIndex(uint256 value) external override onlyOwner {
+    tpi = value;
```

Comment:
You will need to emit events in each setter function, and also in the redeem function.

Will allow things like subgraph to operate (and also good practice so people can track events on chain)

Reply:
[2eba8d9](https://github.com/TempleDAO/temple/pull/760/commits/2eba8d9c319da00aa83f35f11e2f2d2123b34865)